README.md
<p align="center">
<img width="200px" src="http://cloudrail.github.io/img/cloudrail_logo_github.png"/>
</p>

# CloudRail SI for iOS

Integrate Multiple Services With Just One API

<p align="center">
<img width="300px" src="http://cloudrail.github.io/img/cloudrail_si_github.png"/>
</p>

CloudRail is an API integration solution which abstracts multiple APIs from different providers into a single and universal interface.

**Current Interfaces:**
<p align="center">
<img width="800px" src="http://cloudrail.github.io/img/available_interfaces_v3.png"/>
</p>
---
---

[![CocoaPods](https://img.shields.io/cocoapods/v/cloudrail-si-ios-sdk.svg)]()
[![CocoaPods](https://img.shields.io/cocoapods/p/cloudrail-si-ios-sdk.svg)]()
<!--[![CocoaPods](https://img.shields.io/cocoapods/l/cloudrail-si-ios-sdk.svg?maxAge=2592000)]()-->
<!--[![CocoaPods](https://img.shields.io/cocoapods/metrics/doc-percent/cloudrail-si-ios-sdk.svg?maxAge=2592000)]()-->
<!--[![CocoaPods](https://img.shields.io/cocoapods/aw/cloudrail-si-ios-sdk.svg?maxAge=2592000)]()-->
<!--[![CocoaPods](https://img.shields.io/cocoapods/at/cloudrail-si-ios-sdk.svg?maxAge=2592000)]()-->
<!--asdas-->
<!--[![CocoaPods](https://img.shields.io/cocoapods/dw/cloudrail-si-ios-sdk.svg?maxAge=2592000)]()-->
<!--[![CocoaPods](https://img.shields.io/cocoapods/dm/cloudrail-si-ios-sdk.svg?maxAge=2592000)]()-->
<!--[![CocoaPods](https://img.shields.io/cocoapods/dt/cloudrail-si-ios-sdk.svg?maxAge=2592000)]()-->

Full documentation can be found at our [website](https://cloudrail.com/integrations).

Learn more about CloudRail on https://cloudrail.com

---
---

With CloudRail, you can easily integrate external APIs into your application. 
CloudRail provides abstracted interfaces that take several services and then exposes a developer-friendly API that uses common functions between all providers. 
This means that, for example, upload() works in exactly the same way for Dropbox as it does for Google Drive, OneDrive, and other Cloud Storage Services, and getEmail() works similarly the same way across all social networks.

## CocoaPods
Add the pod to your podfile
```
pod 'cloudrail-si-ios-sdk'
```
run
```
pod install
```

## Basic setup
In the Project Navigator, open the source file of your application delegate. Add the import statement at the top of the file, then add the following call to CloudRail in the didFinishLaunching or didFinishLaunchingWithOptions method of your app delegate:
#### Objective-C
```objective-c
#import <CloudRailSI/CloudRailSI.h>

[CRCloudRail setAppKey:@"{Your_License_Key}"];

```

#### Swift
```swift
import CloudrailSI

CRCloudRail.setAppKey("{Your_License_Key}")

```
---


## Current Interfaces:
Interface | Included Services
--- | ---
Cloud Storage | Dropbox, Google Drive, OneDrive, Box, PCloud, Egnyte, OneDrive Business
Business Cloud Storage | Amazon S3, Microsoft Azure, Rackspace, Backblaze
Social Profiles | Facebook, GitHub, Google+, LinkedIn, Slack, Twitter, Windows Live, Yahoo, Instagram, Heroku
Social Interaction | Facebook, Twitter
Payment | PayPal, Stripe
Email | Maljet, Sendgrid, Gmail
SMS | Twilio, Nexmo
Point of Interest | Google Places, Foursquare, Yelp
Video | Youtube, Twitch, Vimeo
Messaging | Facebook Messenger, Telegram, Line, Viber

---
### Cloud Storage Interface:

* Dropbox
* Box
* Google Drive
* Microsoft OneDrive
* PCloud
* Egnyte
* OneDrive Business

#### Features:

* Download files from Cloud Storage.
* Upload files to Cloud Storage.
* Get Meta Data of files, folders and perform all standard operations (copy, move, etc) with them.
* Retrieve user and quota information.
* Generate share links for files and folders.

#### Code Sample - Objective-C
[Full Documentation](https://cloudrail.com/integrations/interfaces/CloudStorage;platformId=ObjectiveC)
```objective-c
id<CRCloudStorageProtocol> service;

//   service = [[CROneDrive alloc] initWithClientId:@"clientIdentifier" clientSecret:@"clientSecret" redirectUri:@"REDIRURL" state:@"CRSTATE"];
//   service = [[CRGoogleDrive alloc] initWithClientId:@"clientIdentifier" clientSecret:@"" redirectUri:@"REDIRURL" state:@"CRSTATE"];
//   service = [[CRBox alloc] initWithClientId:@"clientIdentifier" clientSecret:@"clientSecret" redirectUri:@"REDIRURL" state:@"CRSTATE"];
//   service = [[CRPCloud alloc] initWithClientId:@"clientIdentifier" clientSecret:@"clientSecret" redirectUri:@"REDIRURL" state:@"CRSTATE"];

service = [[CRDropbox alloc] initWithClientId:@"clientIdentifier" clientSecret:@"clientSecret" redirectUri:@"REDIRURL" state:@"CRSTATE"];
NSInputStream * object = [service downloadFileWithPath:@"/mudkip.jpg"];
//READ FROM STREAM
```
#### Code Sample - Swift
[Full Documentation](https://cloudrail.com/integrations/interfaces/CloudStorage;platformId=Swift)
```swift
//let cloudStorage : CloudStorageProtocol = Box.init(clientId: "ClientID", clientSecret: "ClientSecret")
//let cloudStorage : CloudStorageProtocol = GoogleDrive.init(clientId: "ClientID", clientSecret: "", redirectUri: "[redirectUri]", state: "[state]")
//let cloudStorage : CloudStorageProtocol = OneDrive.init(clientId: "ClientID", clientSecret: "ClientSecret")
//let cloudStorage : CloudStorageProtocol = PCloud.init(clientId: "ClientID", clientSecret: "ClientSecret")

let cloudStorage : CloudStorageProtocol = Dropbox.init(clientId: "ClientID", clientSecret: "ClientSecret")
do {
  let inputStream = try cloudStorage.downloadFileWithPath("/TestFolder/Data.csv")
} catch let error{
  print("An error: \(error)")
}
//READ FROM STREAM
```
---
### Business/Bucket Cloud Storage Interface:

* Amazon Web Services S3
* Microsoft Azure
* Rackspace
* Backblaze

#### Features

* Create, delete and list buckets
* Upload files
* Download files
* List files in a bucket and delete files
* Get file metadata (last modified, size, etc.)


#### Code Sample - Objective-C
[Full Documentation](https://cloudrail.com/integrations/interfaces/BusinessCloudStorage;platformId=ObjectiveC)

```objective-c
id<CRBusinessCloudStorageProtocol> service;

// service = [[CRMicrosoftAzure alloc] initWithAccountName:@"[account_name]" accessKey:@"[access_key]"];
// service = [[CRAmazonS3 alloc] initWithAccessKeyId:@"[clientID]" secretAccessKey:@"[client_Secret]" region:@"[region]"];
// service = [[CRRackspace alloc] initWithUsername:@"[username]" apiKey:@"[api_key]" region:@"[region]"];

service = [[CRBackblaze alloc] initWithAccountID:@"[account_init]" appKey:@"[app_key]"];

CRBucket * bucket = [[CRBucket alloc] init];
bucket.name = @"[bucket_name]";
bucket.identifier = @"[bucket_id]";

NSData * data = //data source;
NSInputStream * inputStream = [NSInputStream inputStreamWithData:data];
@try {
  [service uploadFileToBucket:bucket name:@"[file_name]" withStream:inputStream size:data.length];
} @catch (NSException *exception) {
  //handle exception
}
```
#### Code Sample - Swift
[Full Documentation](https://cloudrail.com/integrations/interfaces/BusinessCloudStorage;platformId=Swift)
```swift
//let cloudStorage : BusinessCloudStorageProtocol = Backblaze.init(accountID: "[account_id]", appKey: "[app_key]")
//let cloudStorage : BusinessCloudStorageProtocol = Rackspace.init(username: "[username]", apiKey: "[api_key]", region: "[region]")
//let cloudStorage : BusinessCloudStorageProtocol = MicrosoftAzure.init(accountName: "[account_name]", accessKey: "access_key")
let cloudStorage : BusinessCloudStorageProtocol = AmazonS3.init(accessKeyId: "[access_key]", secretAccessKey: "[secret_key]", region: "[region]")

let bucket : CRBucket = CRBucket.init()
bucket.name = "[bucketName]";
bucket.identifier = "[identifier]"

let path = Bundle.main.path(forResource: "UserData", ofType: "csv")!
let fileAttributes = try! FileManager.default.attributesOfItem(atPath: path)
let fileSize: UInt64 = fileAttributes[FileAttributeKey.size] as! UInt64
let stream = InputStream.init(fileAtPath: path)!

do {
  try cloudStorage.uploadFileToBucket(bucket, name: "[fileName]", stream: stream, size:fileSize)
} catch let error{
  print("An error: \(error)")
}
```
---
### Social Media Profiles Interface:

* Facebook
* Github
* Google Plus
* LinkedIn
* Slack
* Twitter
* Windows Live
* Yahoo
* Instagram
* Heroku

#### Features

* Get profile information, including full names, emails, genders, date of birth, and locales.
* Retrieve profile pictures.
* Login using the Social Network.

#### Code Sample - Objective-C
[Full Documentation](https://cloudrail.com/integrations/interfaces/Profile;platformId=ObjectiveC)

```objective-c
id<CRProfileProtocol> service;

//  service = [[CRGitHub alloc] initWithClientId:@"clientIdentifier" clientSecret:@"clientSecret" redirectUri:@"REDIRURL" state:@"CRSTATE"];
//  service = [[CRInstagram alloc] initWithClientId:@"clientIdentifier" clientSecret:@"clientSecret" redirectUri:@"REDIRURL" state:@"CRSTATE"];
//  service = [[CRSlack alloc] initWithClientId:@"clientIdentifier" clientSecret:@"clientSecret" redirectUri:@"REDIRURL" state:@"CRSTATE"];
//  service = [[CRGooglePlus alloc] initWithClientId:@"clientIdentifier" clientSecret:@"clientSecret" redirectUri:@"REDIRURL" state:@"CRSTATE"];

service = [[CRFacebook alloc] initWithClientId:@"clientIdentifier" clientSecret:@"clientSecret" redirectUri:@"REDIRURL" state:@"CRSTATE"];

NSString * fullName = [service fullName];
```

#### Code Sample - Swift
[Full Documentation](https://cloudrail.com/integrations/interfaces/Profile;platformId=Swift)
```swift
// let profile = GitHub(clientID: "[clientID]", clientSecret: "[clientSecret]")
// let profile = GooglePlus(clientID: "[clientID]", clientSecret: "[clientSecret]")
// let profile = Instagram(clientID: "[clientID]", clientSecret: "[clientSecret]")
// let profile = Slack(clientID: "[clientID]", clientSecret: "[clientSecret]")

let profile = Facebook(clientID: "[clientID]", clientSecret: "[clientSecret]")
do {
  let fullName = try profile.fullName()
} catch let error{
  print("An error: \(error)")
}
```

---

### Social Media Interaction Interface:

* Facebook
* Twitter

#### Features

* Get a list of connections.
* Make a post for the user.


#### Code Sample - Objective-C
[Full Documentation](https://cloudrail.com/integrations/interfaces/Social;platformId=ObjectiveC)

```objective-c
id<CRSocialProtocol> service;

// service = [[CRTwitter alloc] initWithClientIdentifier:@"clientIdentifier" clientSecret:@"clientSecret" redirectUri:@"REDIRURL"];

service = [[CRFacebook alloc] initWithClientId:@"clientIdentifier" clientSecret:@"clientSecret" redirectUri:@"REDIRURL" state:@"CRSTATE"];
[service postUpdateWithContent:@"Using Cloudrail sdk!"];
```
#### Code Sample - Swift
[Full Documentation](https://cloudrail.com/integrations/interfaces/Social;platformId=Swift)

```swift
// let social = Twitter(clientID: "[clientID]", clientSecret: "[clientSecret]")

let social = Facebook(clientID: "[clientID]", clientSecret: "[clientSecret]")
do {
  try social.postUpdateWithContent("CloudRail is awesome!!!")
} catch let error{
  print("An error: \(error)")
}
```
---

### Payment Interface:

* PayPal
* Stripe

#### Features Interface

* Perform charges
* Refund previously made charges
* Manage subscriptions

#### Code Sample - Objective-C
[Full Documentation](https://cloudrail.com/integrations/interfaces/Payment;platformId=ObjectiveC)

```objective-c
id<CRPaymentProtocol> service;

//  service = [[CRPayPal alloc] initWithUseSandbox:YES clientId:@"key" clientSecret:@"secret"];

service = [[CRStripe alloc] initWithSecretKey:@"key"];

SubscriptionPlan * subPlan = [service createSubscriptionPlanWithName:@"Plan name" amount:@2000 currency:@"USD" description:@"description" Longerval:@"day" Longerval_count:@7];

NSLog(@"Sub plan %@", subPlan);
```

#### Code Sample - Swift
[Full Documentation](https://cloudrail.com/integrations/interfaces/Payment;platformId=Swift)

```swift

let payment = PayPal(useSandbox: [true/false], clientId: "[clientID]")
do {
  let subscriptionPlan : CRSubscriptionPlan = try payment.createSubscriptionPlanWithName("My subscription", amount: 500, currency: "USD", description: "myDescription", interval: "week", intervalCount: 2)
} catch let error{
  print("An error: \(error)")
}

```
---

### Email Interface:

* Mailjet
* Sendgrid
* Gmail

#### Features

* Send Email

#### Code Sample - Objective-C
[Full Documentation](https://cloudrail.com/integrations/interfaces/Email;platformId=ObjectiveC)

```objective-c
id<CREmailProtocol> service;

//  service = [[CRMailJet alloc] initWithClientId:@"key" clientSecret:@"secret"];
// service = [[CRGMail alloc] initWithClientID:@"[GMail Client Identifier]" clientSecret:@"" redirectUri:@"com.cloudrail.example:/auth" state:@"someState"];
//[service useAdvancedAuthentication]; //required for Gmail

service = [[CRSendGrid alloc]initWithUsername:@"key" password:@"secret"];

CRAttachment *attachImage = [[CRAttachment alloc] initWithContent:NSInputStream type:"image/jpg" name:"fileName"];
NSMutableArray *attachments = [[NSMutableArray alloc] init];
[attachments addObject:attachImage];

[service sendEmailFromAddress:@"cloudrail@cloudrail.com" fromName:@"Bob" toAddresses:[@[@"foo@gmail.com",@"bar@gmail.com"] mutableCopy] subject:@"Mailjet and SendGrid" textBody:@"The Mailjet and Sendgrid is on cloudrail now!!!" htmlBody:@"" ccAddresses:[@[]mutableCopy] bccAddresses:[@[] mutableCopy], attachments:attachments];
```
#### Code Sample - Swift
[Full Documentation](https://cloudrail.com/integrations/interfaces/Email;platformId=Swift)

```swift

var service: EmailProtocol;

// service = SendGrid(apiKey: "[SendGrid API Key]")
// service = GMail(clientID: "[GMail Client Identifier]",clientSecret: "",redirectUri: "com.cloudrail.example:/auth",state: "someState")
// service.useAdvancedAuthentication() //required for Gmail

service = MailJet(clientID: "[clientID]", clientSecret:"[accountSid]")

let attachImage = CRAttachment()
attachImage.content = InputStream
attachImage.mimeType = "image/jpg"
attachImage.filename = "File.jpg"

var attachments = [CRAttachment]()
attachments.append(attachImage)


do {
  try service.sendEmailFromAddress("info@cloudrail.com", fromName: "CloudRail", toAddresses:NSMutableArray(array:  ["foo@bar.com","bar@foo.com"]), subject: "my subject", textBody: "text body", htmlBody: "Html body", ccAddresses: NSMutableArray(array:  ["foo@bar.com","bar@foo.com"]), bccAddresses: NSMutableArray(array:  ["foo@bar.com","bar@foo.com"])  attachments: attachments)
} catch let error{
  print("An error: \(error)")
}
```
---

### SMS Interface:

* Twilio
* Nexmo
* Twizo

#### Features

* Send SMS

#### Code Sample - Objective-C
[Full Documentation](https://cloudrail.com/integrations/interfaces/SMS;platformId=ObjectiveC)

```objective-c
id<CRSMSProtocol> service;

//service = [[CRNexmo alloc] initWithClientId:@"key" clientSecret:@"secret"];
service = [[CRTwilio alloc] initWithAccountSid:@"key" authToken:@"secret"];

[service sendSmsFromName:@"from Code" toNumber:@"+12323423423" content:@"Testing message"];
```
#### Code Sample - Swift
[Full Documentation](https://cloudrail.com/integrations/interfaces/SMS;platformId=Swift)

```swift
// let sms = Nexmo(accountSid: "[clientID]", authToken: "[authToken]")

let sms = Twilio(accountSid: "[clientID]", authToken: "[authToken]")
do {
  try sms.sendSmsFromName("CloudRail", toNumber: "+491234567890", content: "Hello from CloudRail")
} catch let error{
  print("An error: \(error)")
}
```

---

### Points of Interest Interface:

* Google Places
* Foursquare
* Yelp

#### Features

* Get a list of POIs nearby
* Filter by categories or search term

#### Code Example - Objective-C
[Full Documentation](https://cloudrail.com/integrations/interfaces/PointsOfInterest;platformId=ObjectiveC)

```objective-c
id<CRPointsOfInterestProtocol> service;

//  service = [[CRYelp alloc] initWithConsumerKey:@"key" consumerSecret:@"secret" token:@"token"  tokenSecret:@"tokensecret"];
//  service = [[CRGooglePlaces alloc] initWithApiKey:@"apiKey"];

service = [[CRFoursquare alloc] initWithClientId:key clientSecret:secret];
NSMutableArray<POI*>* pois =  [service nearbyPoisWithLatitude:@49.483927 longitude:@8.473272 radius:@300 searchTerm:[NSNull null] categories:[NSNull null]];

NSLog(@"%@", pois);
```
#### Code Example - Swift
[Full Documentation](https://cloudrail.com/integrations/interfaces/PointsOfInterest;platformId=Swift)

```swift
// let points = GooglePlaces(apiKey: "[secretKey]")

let points = Foursquare(clientID: "[cientID]", clientSecret: "[clientSecret]")
do {
  //Mannheim : 49.483927, 8.473272
  var points = try places.nearbyPoisWithLatitude(49.483927, longitude: 8.473272, radius: 3000, searchTerm: "Restaurant", categories: [])
} catch let error{
  print("An error: \(error)")
}
```
---

### Video Interface:

* Youtube
* Twitch
* Vimeo

#### Features

* Search for videos
* Upload videos
* Get a list of videos for a channel
* Get channel details
* Get your own channel details
* Get video details 

#### Code Example - Objective-C
[Full Documentation](https://cloudrail.com/integrations/interfaces/Video;platformId=ObjectiveC)

```objective-c
id<CRVideoProtocol> service;

//  service = [[CRTwitch alloc] initWithClientId:@"[Twitch Client Identifier]" clientSecret:@"[Twitch Client Secret]"];
//  service = [[CRVimeo alloc] initWithClientId:@"[Vimeo Client Identifier]" clientSecret:@"[Vimeo Client Secret]"];

service = [[CRYouTube alloc] initWithClientId:@"[YouTube Client Identifier]" clientSecret:@"" redirectUri:@"com.cloudrail.example:/auth" state:@"someState"];
[service useAdvancedAuthentication]; //required for Youtube


NSMutableArray<CRVideoMetaData *> * searchResult = [service searchVideosWithQueryQuery:@"CloudRail" offset:[NSNumber numberWithInt: 0] limit:[NSNumber numberWithInt: 12]];

NSLog(@"%@", searchResult);

CRVideoMetaData * uploadResult = [service uploadVideoWithTitleTitle:@"HowTo: Setup CloudRail" 
        description:@"Video about Setting up CloudRail" 
        stream:NSInputStream 
        size:[NSNumber numberWithInt: 1448576] 
        channelId:@"70207321" 
        mimeType:@"video/mp4"
]; //ChannelId optional for youtube

NSLog(@"%@", uploadResult);


```
#### Code Example - Swift
[Full Documentation](https://cloudrail.com/integrations/interfaces/Video;platformId=Swift)

```swift
var service: VideoProtocol

// service = Twitch(clientId: "[Twitch Client Identifier]",clientSecret: "[Twitch Client Secret]")
// service = Vimeo(clientId: "[Vimeo Client Identifier]",clientSecret: "[Vimeo Client Secret]")

service = YouTube(clientId: "[YouTube Client Identifier]",clientSecret: "",redirectUri: "com.cloudrail.example:/auth",state: "someState")
service.useAdvancedAuthentication(); //required for Youtube

do {
   
   let searchResult = service.searchVideos(query: "CloudRail",offset: 0,limit: 12)
   print(searchResult)
   
   let uploadResult = service.uploadVideo(title: "HowTo: Setup CloudRail", description: "Video about Setting up CloudRail", stream: InputStream, size: 1448576, channelId: "70207321", mimeType: "video/mp4") //Channel id optional for youtube
   
   print(uploadResult)
   
} catch let error{
  print("An error: \(error)")
}
```
---
### Messaging Interface:

* FacebookMessenger
* Telegram
* Line
* Viber

#### Features

* Send text messages
* Send files, images, videos and audios
* Parse a message received on your webhook
* Download the content of an attachment sent to your webhook

#### Code Example - Objective-C
[Full Documentation](https://cloudrail.com/integrations/interfaces/Messaging;platformId=ObjectiveC)

```objective-c
id<CRMessagingProtocol> service;

//  service = [[CRViber alloc] initWithBotToken:@"[Bot Token]" webhook:@"[Webhook URL]" botName:@"[Bot Name]"];
//  service = [[CRTelegram alloc] initWithBotToken:@"[Bot Token]" webhook:@"[Webhook URL]"];
//  service = [[CRLine alloc] initWithBotToken:@"[Bot Token]"];

service =  [[CRFacebookMessenger alloc] initWithBotToken:@"[Bot Token]"];

CRMessage * result = [service sendMessageWithReceiverIdReceiverId:@"12123242" 
        message:@"It's so easy to send message via CloudRail"
];

NSLog(@"%@", result);

```
#### Code Example - Swift
[Full Documentation](https://cloudrail.com/integrations/interfaces/Messaging;platformId=Swift)

```swift
var service: MessagingProtocol

// service = Viber(botToken: "[Bot Token]",webhook: "[Webhook URL]",botName: "[Bot Name]")
// service = Telegram(botToken: "[Bot Token]",webhook: "[Webhook URL]")
// service = Line(botToken: "[Bot Token]")

service = FacebookMessenger(botToken: "[Bot Token]")

do {
   
   var result = service.sendMessageWithReceiverId(
    receiverId: "12123242",
    message: "It's so easy to send message via CloudRail")
   
   print(result)
   
} catch let error{
  print("An error: \(error)")
}
```


---

More interfaces are coming soon.

## Advantages of Using CloudRail

* Consistent Interfaces: As functions work the same across all services, you can perform tasks between services simply.

* Easy Authentication: CloudRail includes easy ways to authenticate, to remove one of the biggest hassles of coding for external APIs.

* Switch services instantly: One line of code is needed to set up the service you are using. Changing which service is as simple as changing the name to the one you wish to use.

* Simple Documentation: There is no searching around Stack Overflow for the answer. The CloudRail documentation at https://cloudrail.com/integrations is regularly updated, clean, and simple to use.

* No Maintenance Times: The CloudRail Libraries are updated when a provider changes their API.

* Direct Data: Everything happens directly in the Library. No data ever passes a CloudRail server.

## Integrate With Cocoapods (Swift & Objective-C)

CloudRail-SI-iOS is available through [CocoaPods](http://cocoapods.org). To install
it, simply add the following line to your Podfile (remember to put the `use_frameworks!` flag on top of the `Podfile`):
```ruby
pod "cloudrail-si-ios-sdk"
```
Run `Pod install` again.
### Swift

* Import the module with `import CloudrailSI` on swift files (Classes, Protocols. etc...).


### Objective-C

* Import the framework on the desired class with `#import <CloudRailSI/CloudRailSI.h>`.

## Integrate Into Projects Without Cocoapods

### Swift (use Swift interface)

1. Drag an drop the Framework file to the __"Embedded Binaries"__ of the iOS project, check __"copy files"__ if needed.
2. You can import the module with `import CloudrailSI` on swift files (Classes, Protocols. etc...).


### Swift (use ObjC interface)

1. add a new __Objective-C__ File (any file will do) to your project and Xcode will prompt if you want to configure your project with a bridging header (`PROJECTNAME-Bridging-Header.h`), press __YES__ on the prompt. (In case you can not add a bridging header use [this](http://www.learnswiftonline.com/getting-started/adding-swift-bridging-header/) guide)
2. Drag an drop the Framework file to the __"Embedded Binaries"__ of the iOS project, check __"copy files"__ if needed.
3. Xcode will generate and configure the file for you, on the file you have to import ( in a Objective-C way) with `#import <CloudRailSI/CloudRailSI.h>`.

### Objective-C

Simply drag an drop the Framework file to the __"Embedded Binaries"__ of the iOS project, check __"copy files"__ if needed. Import the framework on the desired class with `#import <CloudRailSI/CloudRailSI.h>`, and have fun!

# Start implementing

Now that you are all set up, you can learn how to use CloudRail by heading over to [[Usage]] or [[Usage-(Swift)]]
## Other Code Samples
### Swift

```swift
//let cloudStorage : CloudStorageProtocol = Box.init(clientId: "ClientID", clientSecret: "ClientSecret")
//let cloudStorage : CloudStorageProtocol = GoogleDrive.init(clientId: "ClientID", clientSecret: "ClientSecret")
//let cloudStorage : CloudStorageProtocol = OneDrive.init(clientId: "ClientID", clientSecret: "ClientSecret")

CHANGELOG.md
# Change Log

## [5.13.0](https://github.com/CloudRail/cloudrail-si-ios-sdk/tree/5.13.0) (2018-09-23)
[Full Changelog](https://github.com/CloudRail/cloudrail-si-ios-sdk/compare/5.12.1...5.13.0)

* Updated with new compiler. (Swift 4.2)
* Bug fixes related to rootId in GoogleDrive
* Bug fixes in backblaze

## [5.12.1](https://github.com/CloudRail/cloudrail-si-ios-sdk/tree/5.12.1) (2018-07-11)
[Full Changelog](https://github.com/CloudRail/cloudrail-si-ios-sdk/compare/5.12.0...5.12.1)

* Fixed bug that freezes during advanced authentication
* Minor Bug fixes

## [5.12.0](https://github.com/CloudRail/cloudrail-si-ios-sdk/tree/5.12.0) (2018-07-05)
[Full Changelog](https://github.com/CloudRail/cloudrail-si-ios-sdk/compare/5.11.3...5.12.0)

* Updated to swift 4.1.2
* Added AuthenticatingDelegate
* Added UploadDidCancel method to UploadDelegate
* Minor Bug fixes

## [5.11.3](https://github.com/CloudRail/cloudrail-si-ios-sdk/tree/5.11.3) (2018-05-22)
[Full Changelog](https://github.com/CloudRail/cloudrail-si-ios-sdk/compare/5.11.2...5.11.3)

* OneDriveBusiness: Fix missing url encoding of the client secret on refreshing tokens
* MicrosoftAzure: Fix an issue in the advanced request function
* Fixed bug when performing authentication in background thread

## [5.11.2](https://github.com/CloudRail/cloudrail-si-ios-sdk/tree/5.11.2) (2018-05-11)
[Full Changelog](https://github.com/CloudRail/cloudrail-si-ios-sdk/compare/5.11.1...5.11.2)

* Fixed bug that UI updates where not executed on main thread during authentication
* Minor bug fixes & stability improvement
* Removed unused

## [5.11.1](https://github.com/CloudRail/cloudrail-si-ios-sdk/tree/5.11.1) (2018-04-24)
[Full Changelog](https://github.com/CloudRail/cloudrail-si-ios-sdk/compare/5.11.0...5.11.1)

* Fixed bug where backslash file path where not encoded properly in GoogleDrive.
* OneDrive: Attach missing state parameter to authentication url
* Facebook: Removed user_about_me scope which shows description. This is removed from Facebook
* Google Cloud: Added object name encoding to google cloud.

## [5.11.0](https://github.com/CloudRail/cloudrail-si-ios-sdk/tree/5.11.0) (2018-04-10)
[Full Changelog](https://github.com/CloudRail/cloudrail-si-ios-sdk/compare/5.10.0...5.11.0)

* Updated to **Swift 4.1**
* Fixed bug in **fileExistsAtPath:** method in GoogleDrive.
* Added AdvancedRequest feature in Yelp.
* Added new method **listFilesInBucket:withPrefix:** to BusinessCloudStorage interface.

## [5.10.0](https://github.com/CloudRail/cloudrail-si-ios-sdk/tree/5.10.0) (2018-03-27)
[Full Changelog](https://github.com/CloudRail/cloudrail-si-ios-sdk/compare/5.9.13...5.10.0)

* Minor fix in AmazonS3, GoogleCloudPlatform, FacebookMessenger
* Fix in uploadFileWithContentModified for all CloudStorage Interfaces.
* Updated Yelp to use API v3.

## [5.9.14](https://github.com/CloudRail/cloudrail-si-ios-sdk/tree/5.9.14) (2018-02-28)
[Full Changelog](https://github.com/CloudRail/cloudrail-si-ios-sdk/compare/5.9.13...5.9.14)

* Minor changes to OneDrive, OneDriveBusiness, GoogleDrive
* Fixed Unknown Issue caused by wrong allocation
* Fixed apostrophe issue in GoogleDrive
* Fixed Crash on OneDrive
* Updated CRDate formats

## [5.9.13](https://github.com/CloudRail/cloudrail-si-ios-sdk/tree/5.9.13) (2018-01-23)
[Full Changelog](https://github.com/CloudRail/cloudrail-si-ios-sdk/compare/5.9.12...5.9.13)

* Added missing protocol inheritance in SocialProtocol Swift
* Added uploadedWithModifiedDate to CRCloudStorage
* Updated CloudMetaData properties

## [5.9.12](https://github.com/CloudRail/cloudrail-si-ios-sdk/tree/5.9.12) (2018-01-23)
[Full Changelog](https://github.com/CloudRail/cloudrail-si-ios-sdk/compare/5.9.11...5.9.12)

* PCloud added as a new integration in the Cloud Storage interface
* Fixed Google Drive path issue (when it contains backslashes)

## [5.9.11](https://github.com/CloudRail/cloudrail-si-ios-sdk/tree/5.9.11) (2017-12-22)
[Full Changelog](https://github.com/CloudRail/cloudrail-si-ios-sdk/compare/5.9.10...5.9.11)

* Fixed Issue where Google services throw exception before authentication

## [5.9.7](https://github.com/CloudRail/cloudrail-si-ios-sdk/tree/5.9.7) (2017-11-28)
[Full Changelog](https://github.com/CloudRail/cloudrail-si-ios-sdk/compare/5.9.6...5.9.7)

* Hot fix - Connection
* Amazon S3 fix

## [5.9.6](https://github.com/CloudRail/cloudrail-si-ios-sdk/tree/5.9.6) (2017-11-23)
[Full Changelog](https://github.com/CloudRail/cloudrail-si-ios-sdk/compare/5.9.5...5.9.6)

* Added missing files (messaging files)
* Fixed CloudStorage issues to limit files
* Added Category to POI

## [5.9.5](https://github.com/CloudRail/cloudrail-si-ios-sdk/tree/5.9.5) (2017-11-22)
[Full Changelog](https://github.com/CloudRail/cloudrail-si-ios-sdk/compare/5.9.4...5.9.5)

* Updated to Swift 3.2.2 (Using XCode 9.1)

## [5.9.4](https://github.com/CloudRail/cloudrail-si-ios-sdk/tree/5.9.4) (2017-11-07)
[Full Changelog](https://github.com/CloudRail/cloudrail-si-ios-sdk/compare/5.9.3...5.9.4)

* Added Advance Request to Messaging Services
* Added Send Caurosel Method to Messaging Interface
* Fixed the Done button on the WebView (Changed to WKWebView)

## [5.9.3](https://github.com/CloudRail/cloudrail-si-ios-sdk/tree/5.9.3) (2017-10-17)
[Full Changelog](https://github.com/CloudRail/cloudrail-si-ios-sdk/compare/5.9.2...5.9.3)

* Added New Messaging Interfaces (Facebook / Viber)
* Fixed Amazon S3 pagination issue
* Fixed for CloudStorage (OneDrive, Google Drive, Box and Dropbox)
* Fixed OneDrive Business

## [5.9.2](https://github.com/CloudRail/cloudrail-si-ios-sdk/tree/5.9.2) (2017-10-09)
[Full Changelog](https://github.com/CloudRail/cloudrail-si-ios-sdk/compare/5.9.1...5.9.2)

* Updated Messaging Interface
* Fixed timestamp issue
* Fixed missing constructor for OneDrive / Gmail

## [5.9.1](https://github.com/CloudRail/cloudrail-si-ios-sdk/tree/5.9.1) (2017-10-02)
[Full Changelog](https://github.com/CloudRail/cloudrail-si-ios-sdk/compare/5.9.0...5.9.1)


Example/CloudRail-SI-iOS/PoiViewController.m
//
//  PoiViewController.m
//  CloudRail_ServiceCode
//
//  Created by Felipe Cesar on 22/06/16.
//  Copyright © 2016 CloudRail. All rights reserved.
//

#import "PoiViewController.h"

@implementation PoiViewController

-(void)viewDidLoad{
  
  
  if ([self.serviceName isEqualToString:@"Yelp"]) {
    self.service = [[CRYelp alloc] initWithConsumerKey:@"6aiXfOwD3wcHh1VJnDdDow" consumerSecret:@"Hr_wv-72pm7CJ6rcNMxKdiKkNlE" token:@"Kc4hsHaW1NsL2CYwWuBBZa32AXcTqr3l"  tokenSecret:@"9P8T6eYYllj1qyrdCLhgERNHloU"];
  } else if([self.serviceName isEqualToString:@"GooglePlaces"]){
    self.service = [[CRGooglePlaces alloc] initWithApiKey:@"AIzaSyCMSQyKrvt-Wlauh0k5zP8J7ONcnL6IQT8"];
    
  } else if ([self.serviceName isEqualToString:@"Foursquare"]){
    self.service = [[CRFoursquare alloc] initWithClientId:@"2HVTEYZ3LKKRPNRUFVOGTVIF0WLL0XMUJPZQS0QEB1MAB12P" clientSecret:@"PYFJIRODDMQL5N1WTZPBTTR5EMC2W2RJVIX0H5KAQBI43SXM"];
  }
}

- (IBAction)firstTest:(id)sender {
  //FIX the generator!!
  //MAnnheim : 49.483927, 8.473272
  NSMutableArray<CRPOI*> * poi =  [self.service nearbyPoisWithLatitude:@49.483927 longitude:@8.473272 radius:@300 searchTerm:nil categories:nil];
  NSLog(@"%@", poi);
}
@end
